duplex_http_call.go

```diff
@@ -52,19 +53,26 @@ type duplexHTTPCall struct {
 func newDuplexHTTPCall(
 	ctx context.Context,
 	httpClient HTTPClient,
-	url string,
+	url *url.URL,
 	spec Spec,
 	header http.Header,
 ) *duplexHTTPCall {
+	// ensure we make a copy of the url before we pass along to the
+	// Request. This ensures if a transport out of our control wants
+	// to mutate the req.URL, we don't feel the effects of it.
+	url = cloneURL(url)
```

Comment:
This is erroring on the side of caution since we can't really control what an http.RoundTripper may do, and I'm not sure if it's expected that a RoundTripper is not allowed to mutate this, or is expected to make it's own copy if it were?

Doing this at least mirrors the safety we had when we passed along a string and it was parsed every time.

Reply:
RoundTripper implementations aren't supposed to mutate the URL:

> // RoundTrip should not modify the request, except for
	// consuming and closing the Request's Body. RoundTrip may
	// read fields of the request in a separate goroutine. Callers
	// should not mutate or reuse the request until the Response's
	// Body has been closed.

In practice, I don't think many people read the docs very carefully so this extra safety is a good idea.